### PR TITLE
Refs #29850 -- Added exclusion support to window frames.

### PR DESCRIPTION
**Pull request by sarahboyce**
Ticket: https://code.djangoproject.com/ticket/29850

(It was easier to cherry-pick the remaining commit onto a new branch to resolve conflicts - previous PR: https://github.com/django/django/pull/17401)

### CODE REVIEW

**Comment by sarahboyce** [comment]
buildbot, test on oracle.

File: docs/releases/5.1.txt [hunk 1]
```diff
@@ -174,6 +174,11 @@ Models
 * :class:`~django.db.models.expressions.RowRange` now accepts positive integers
   for the ``start`` argument and negative integers for the ``end`` argument.
 
+* The new ``exclusion`` argument of
+  :class:`~django.db.models.expressions.RowRange` and
+  :class:`~django.db.models.expressions.ValueRange` adds an exclusion clause to
+  the generated SQL when set to a  for supported databases.
```
Comment (felixxm) [inline docs/releases/5.1.txt:180]: Maybe
```suggestion
  :class:`~django.db.models.expressions.ValueRange` allows excluding rows,
  groups, and ties from the window frames.
```

File: docs/ref/models/expressions.txt [hunk 2]
```diff
@@ -976,7 +997,8 @@ released between twelve months before and twelve months after each movie:
 .. versionchanged:: 5.1
 
     Support for positive integer ``start`` and negative integer ``end`` was
-    added for ``RowRange``.
+    added for ``RowRange``. ``exclusion`` was added to ``RowRange`` and
```
Comment (felixxm) [inline docs/ref/models/expressions.txt:1000]:
```suggestion
    added for ``RowRange``. The ``exclusion`` argument was added to ``RowRange`` and
```

File: docs/ref/models/expressions.txt [hunk 3]
```diff
@@ -911,6 +911,27 @@ Both classes return SQL with the template:
 
     %(frame_type)s BETWEEN %(start)s AND %(end)s
 
+
```
Comment (felixxm) [inline docs/ref/models/expressions.txt:914]:
```suggestion
```

File: django/db/models/expressions.py [hunk 4]
```diff
@@ -1848,6 +1849,13 @@ def get_group_by_cols(self):
         return group_by_cols
 
 
+class Exclusion(Enum):
+    CURRENT_ROW = "CURRENT ROW"
+    GROUP = "GROUP"
+    TIES = "TIES"
+    NO_OTHERS = "NO OTHERS"
```
Comment (felixxm) [inline django/db/models/expressions.py:1856]: Add `__repr__` consistent with other enums:
```python
    # A similar format was proposed for Python 3.10.
    def __repr__(self):
        return f"{self.__class__.__qualname__}.{self._name_}"

```

File: django/db/models/expressions.py [hunk 5]
```diff
@@ -1859,23 +1867,37 @@ class WindowFrame(Expression):
...
         start, end = self.window_frame_start_end(
             connection, self.start.value, self.end.value
         )
+        window_frame_template_parts = [self.template]
+        if self.exclusion is not None:
+            exclusion_as_sql = self.get_exclusion()
+            window_frame_template_parts.append(exclusion_as_sql)
         return (
-            self.template
+            " ".join(window_frame_template_parts)
```
Comment (felixxm) [inline django/db/models/expressions.py:1900]: What do you think about adding `exclude` to the `self.template` instead?
`template = "%(frame_type)s BETWEEN %(start)s AND %(end)s%(exclude)s"`
```
def get_exclusion(self):
    ...
    if isinstance(self.exclusion, Exclusion):
        return f" EXCLUDE {self.exclusion.value}"
```

**Comment by felixxm** [comment]
@sarahboyce Thanks :+1: I left small comments.

File: django/db/models/expressions.py [hunk 6]
```diff
@@ -1857,29 +1868,43 @@ class WindowFrame(Expression):
...
     def get_source_expressions(self):
         return [self.start, self.end]
 
+    def get_exclusion(self):
+        if not connection.features.supports_frame_exclusion:
+            raise NotSupportedError(
+                "This backend does not support window frame exclusions."
+            )
+        if self.exclusion is None:
+            return ""
```
Comment (felixxm) [inline django/db/models/expressions.py:1890]: Should not we swap these `if`s?

Comment (sarahboyce) [reply django/db/models/expressions.py:1890]: Ah yes 🤦‍♀️

File: django/db/models/expressions.py [hunk 7]
```diff
@@ -1857,29 +1868,43 @@ class WindowFrame(Expression):
...
+    def get_exclusion(self):
+        if self.exclusion is None:
+            return ""
+        if not connection.features.supports_frame_exclusion:
+            raise NotSupportedError(
+                "This backend does not support window frame exclusions."
+            )
+        if isinstance(self.exclusion, Exclusion):
+            return f" EXCLUDE {self.exclusion.value}"
+        raise ValueError("exclusion must be a valid Exclusion value.")
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/db/models/expressions.py:1893]: For example:
```python
    def __init__(self, start=None, end=None, exclusion=None):
        ...
        if not isinstance(exclusion, (NoneType, Exclusion)):
            raise TypeError(
                "{self.__class__.__qualname__}.exclusion must be an Exclusion instance."
            )
        if exclusion is not None and not connection.features.supports_frame_exclusion:
            raise NotSupportedError(...)
```

[the remark this answers is not in the document]
Comment (sarahboyce) [reply django/db/models/expressions.py]: I move the `NotSupportedError` into the `as_sql()` as I don't won't to add a database connection dependency for `repr` (which would mean I would have to update `NonQueryWindowTests.test_window_frame_repr`)

File: django/db/models/expressions.py [hunk 8]
```diff
@@ -1848,6 +1849,16 @@ def get_group_by_cols(self):
         return group_by_cols
 
 
+class Exclusion(Enum):
```
Comment (felixxm) [inline django/db/models/expressions.py:1852]: `Exclusion` can be misleading, especially as its importable from `django.db.models`. I will rename it to the `WindowFrameExclusion`.

**Comment by felixxm** [comment]
@sarahboyce Thanks :+1: I pushed final edits.

**Pull request merged**